File: solutions/kotlin/phone-number/1/src/main/kotlin/PhoneNumber.kt
class PhoneNumber(txt: String?) {
    private var res: String?

    init {
        res = txt?.filter { it.isDigit() }.toString()

        require(res?.length!! in 10..11)
        if (res?.length == 11) {
            require(res?.get(0) == '1')
            res = res?.substring(1)
        }
        require(res?.get(0)?.digitToInt() in 2..9)
        require(res?.get(3)?.digitToInt() in 2..9)
    }

    val number: String? = res

}

File: solutions/scala/phone-number/1/src/main/scala/PhoneNumber.scala
object PhoneNumber {
    def clean(str: String): Option[String] = {
        val cleaned = str
        .stripPrefix("+1")
        .stripPrefix("1")
        .filter(_.isDigit)
        .zipWithIndex
        .filter(x => {
            if (x._2 == 0 || x._2 == 3) ('2' to '9').contains(x._1)
            else ('0' to '9').contains(x._1)
        })
        
        cleaned.length match {
            case 10 => Some(cleaned.map(_._1).mkString)
            case _ => None
        }
    }
}